<h1 style="color:#89986D; font-family:lobster; text-align:center; font-size:255%">Matríz Fundamental</h1>
<p style="text-align:center; font-style:italic; font-size:115%">Cervantes Silva Ricardo</p>
<p style="font-family: arial; text-align:justify; font-size:150%">En este archivo, se resolveran, tanto análiticamente como por somulación, los siguientes cuestionamientos:</p>
<ul style="font-family: arial; text-align:justify; font-size:150%">
    <li>¿Cuál es el número promedio de tiradas necesarias para terminar el juego de serpientes y escaleras en el tablero de la imagen adjunta?</li>
    <li>¿Cuál es la probabilidad de que un ratón, iniciando  en la casilla 0, alcance la comida?</li>
</ul>
    <h1 style="color:#767F9E; font-family:lobster; text-align:center; font-size:200%">Serpientes y escaleras</h1>
    <ul style="font-family: arial; text-align:justify; font-size:150%">
        <li><b>Descipción del tablero</b>: El tablero tiene 20 casillas.
        <ul>
            <li><b>Inicio:</b> casilla 1.</li>
            <li><b>Final:</b> casilla 20.</li>
            <li><b>Escaleras:</b><br>\(3 \rightarrow 11\)<br>\(18 \rightarrow 19\)
            </li>
            <li><b>Serpientes:</b><br>\(13 \rightarrow 4\)<br>\(17 \rightarrow 9\)
</li>
<li>Se utiliza un dado justo de 6 caras.</li>
        <li><b>Regla Utilizada:</b> Si una tirada excede la casilla 20, el jugador permanece en la misma casilla.</li></ul></li><br>
        <li><b>Modelo matemático:</b> Sea \(X_{n}\) la posición del jugador después de la tirada \(n\), el proceso \(\{ X_{n} \}_{n \geq 0}\) es una cadena de Markov, donde la casilla 20 es absorbente.</li><br>
        <li><b>Matriz de transición:</b> Las probabilidades de transición se construyen considerando:<br>
            <ul>
                <li>Resultado del dado.</li>
                <li>Movimiento normal.</li>
                <li>Aplicación automática de serpientes o escaleras.</li>
            El siguiente código en Python se utiliza para construir la matiz de transición.</li>
    </ul>

In [5]:
import sympy as sp
# Número de estados
n = 20
# Serpientes y escaleras
jump = {
    3: 11,
    18: 19,
    13: 4,
    17: 9
}
# Matriz de transición
P = sp.zeros(n, n)
for s in range(1, 21):
    # Estado absorbente
    if s == 20:
        P[s-1, s-1] = 1
        continue
    for d in range(1, 7):
        nxt = s + d
        # Si se excede 20, permanece
        if nxt > 20:
            nxt = s
        # Aplicar serpiente o escalera
        if nxt in jump:
            nxt = jump[nxt]

        P[s-1, nxt-1] += sp.Rational(1, 6)
P


Matrix([
[0, 1/6, 0, 1/6, 1/6, 1/6, 1/6,   0,   0,   0, 1/6,   0, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6, 1/6, 1/6, 1/6, 1/6,   0,   0, 1/6,   0, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6, 1/6, 1/6, 1/6, 1/6, 1/6,   0,   0,   0, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0,   0, 1/6, 1/6, 1/6, 1/6, 1/6, 1/6,   0,   0, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0,   0,   0, 1/6, 1/6, 1/6, 1/6, 1/6, 1/6,   0, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0,   0,   0,   0, 1/6, 1/6, 1/6, 1/6, 1/6, 1/6, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6,   0,   0,   0, 1/6, 1/6, 1/6, 1/6, 1/6, 0,   0,   0,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6,   0,   0,   0,   0, 1/6, 1/6, 1/6, 1/6, 0, 1/6,   0,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6,   0,   0,   0,   0,   0, 1/6, 1/6, 1/6, 0, 1/6, 1/6,   0, 0, 0,   0,   0],
[0,   0, 0, 1/6,   0,   0,   0,   0,   0,   0, 1/6, 1/6, 0, 1/6, 1/6, 1/6, 0, 0,   0,   0],
[0,   0, 0, 1/6,   0,   0,   0,   0, 1/6,   0,   0, 1/6, 0, 1/6, 1/6, 1

<ul style="font-family: arial; text-align:justify; font-size:150%">
   <li><b>Forma canónica:</b> La matriz se separa como:
\[
P =
\begin{pmatrix}
Q & R \\
0 & I
\end{pmatrix}
\]
       Donde:
            <ul>
                <li>(\(Q\)): transiciones entre estados transitorios..</li>
                <li>(\(R\)): transiciones hacia estados absorbentes.</li></ul>
       <li><b>Matriz fundamental:</b> La matriz fundamental es:
\[N = (I - Q)^{-1}\]
La suma de cada fila de (\(N\)) proporciona el número esperado de pasos antes de la absorción.</li>
    <li><b>Cálculo analítico:</b></li></ul>

In [6]:
# Submatriz Q
Q = P[:19, :19]

# Identidad
I = sp.eye(19)

# Matriz fundamental
N = (I - Q).inv()

# Vector columna de unos
ones = sp.ones(19, 1)

# Tiempos esperados
T = N * ones

# Tiempo esperado iniciando en la casilla 1
esperanza = sp.simplify(T[0])

print("Valor exacto:")
print(esperanza)

print("Valor decimal:")
print(float(esperanza))


Valor exacto:
95699741/7970658
Valor decimal:
12.00650448181317


<ul style="font-family: arial; text-align:justify; font-size:150%">
    <li><b>Resultado analítico:</b> El número esperado de tiradas para terminar el juego es:
\[E[T] = \frac{95699741}{7970658}\]
Aproximadamente:
\[E[T] \approx 12.0065\]
Por lo tanto: El jugador necesita aproximadamente 12 tiradas para terminar el juego.</li></ul>
<p style="font-family: arial; text-align:justify; font-size:150%">El código que se presenta a continuación, es la simulación correspondiente con ese problema.</p>

In [8]:
import random
import numpy as np

jump = { 3: 11, 18: 19, 13: 4, 17: 9}

def jugar():
    pos = 1
    tiros = 0
    while pos != 20:
        d = random.randint(1, 6)
        nxt = pos + d
        if nxt > 20:
            nxt = pos
        if nxt in jump:
            nxt = jump[nxt]
        pos = nxt
        tiros += 1
    return tiros
    
Nsim = 100000
resultados = [jugar() for _ in range(Nsim)]
print("Promedio simulado:")
print(np.mean(resultados))



Promedio simulado:
12.04515


<p style="font-family: arial; text-align:justify; font-size:150%"><b>Conclusión:</b><br>Como la simulación se aproxima al valor obtenido en la solución analítica, queda validado el modelo de cadena de Markov.</p>

<h1 style="color:#767F9E; font-family:lobster; text-align:center; font-size:200%">Problema del ratón y la comida</h1>
<ul style="font-family: arial; text-align:justify; font-size:150%">
        <li><b>Descipción:</b> El ratón inicia en la casilla 0.
        <ul>
            <li>La casilla 7 contiene comida.</li>
            <li>La casilla 8 contiene una descarga eléctrica.</li>
        </ul>
        El ratón se mueve aleatoriamente hacia cualquiera de las casillas vecinas disponibles con igual probabilidad. Se desea calcular:
\[P(\text{llegar a 7 antes que 8})\]
        </li><br>
        <li><b>Modelo probabilístico:</b> Definimos:
\[p_i = P(\text{alcanzar 7 antes que 8}\mid X_0=i)\]
Condiciones de frontera:
\[p_7 = 1\]
\[p_8 = 0\]</li><br>
        <li><b>Ecuaciónes del sistema:</b> Obtenidas a partir del diagrama<br>
            <ul>
                <li>Estado 0 \[p_0 = \frac{p_1 + p_2}{2}\]</li>
                <li>Estado 1 \[p_1 = \frac{p_0 + p_3}{2}\]</li>
                <li>Estado 2 \[p_2 = \frac{p_0 + p_3 + p_8}{3}\]
                    Como: \[p_8 = 0\]
                    entonces: \[p_2 = \frac{p_0 + p_3}{3}\]</li>
                <li>Estado 3 \[p_3 = \frac{p_1 + p_2 + p_4 + p_5}{4}\]</li>
                <li>Estado 4 \[p_4 = \frac{p_3 + p_6}{2}\]</li>
                <li>Estado 5 \[p_5 = \frac{p_3 + p_6 + p_8}{3}\]
                    Como: \[p_8 = 0\] 
                    entonces: \[p_5 = \frac{p_3 + p_6}{3}\]</li>
                <li>Estado 6 \[p_6 = \frac{p_4 + p_5 + p_7}{3}\]
                    Como: \[p_7 = 1\]
                    entonces: \[p_6 = \frac{p_4 + p_5 + 1}{3}\]</li>
            Para solucionar el sistema de ecuaciónes y obtener la solución analítica, se utilizó el siguiente programa en Python:</li>
    </ul>

In [9]:
import sympy as sp

p0, p1, p2, p3, p4, p5, p6 = sp.symbols(
    'p0 p1 p2 p3 p4 p5 p6'
)

# Sistema de ecuaciones
eqs = [
    sp.Eq(p0, (p1 + p2)/2),
    sp.Eq(p1, (p0 + p3)/2),
    sp.Eq(p2, (p0 + p3)/3),
    sp.Eq(p3, (p1 + p2 + p4 + p5)/4),
    sp.Eq(p4, (p3 + p6)/2),
    sp.Eq(p5, (p3 + p6)/3),
    sp.Eq(p6, (p4 + p5 + 1)/3)
]
sol = sp.solve(eqs, [p0, p1, p2, p3, p4, p5, p6])
sol

{p0: 25/129,
 p1: 10/43,
 p2: 20/129,
 p3: 35/129,
 p4: 18/43,
 p5: 12/43,
 p6: 73/129}

<ul style="font-family: arial; text-align:justify; font-size:150%">
    <li><b>Resultado analítico:</b> La probabilidad buscada es: \[p_0 = \frac{25}{129} \] Aproximadamente: \[p_0 \approx 0.1938\]
</li>
<li><b>Interpretación:</b> La probabilidad de que el ratón alcance la comida antes de la descarga eléctrica es aproximadamente: \(19.38%\)</li></ul>
<p style="font-family: arial; text-align:justify; font-size:150%">El código que se presenta a continuación, es la simulación correspondiente con ese problema.</p>

In [11]:
import random
import numpy as np
# Grafo de conexiones
vecinos = {
    0: [1, 2],
    1: [0, 3],
    2: [0, 3, 8],
    3: [1, 2, 4, 5],
    4: [3, 6],
    5: [3, 6, 8],
    6: [4, 5, 7],
    7: [],
    8: []
}

def experimento():
    estado = 0

    while estado not in [7, 8]:
        estado = random.choice(vecinos[estado])
    return estado == 7
    
Nsim = 100000
resultados = [experimento() for _ in range(Nsim)]
print("Probabilidad simulada:")
print(np.mean(resultados))

Probabilidad simulada:
0.19548


<p style="font-family: arial; text-align:justify; font-size:150%"><b>Conclusión:</b><br>Aunque el resultado no es 100% igual al calculado en la solución análitica, podemos obersvar que el error (algo muy común al relizar cálculos en computadora) es muy bajo, con lo cual, podemos concluir que el resultado esperado en la simulación si corresponde con lo calculado analíticamente.</p>